src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// CA2016: Forward CancellationToken to invocations.
+    /// 
+    /// Conditions for positive cases:
+    ///     - The containing method signature receives a ct parameter. It can be a method, a nested method, an action or a func.
+    ///     - The invocation method is not receiving a ct argument, and...
+    ///     - The invocation method either:
+    ///         - Has no overloads but its current signature receives an optional ct=default, currently implicit, or...
+    ///         - Has a method overload with the exact same arguments in the same order, plus one ct parameter at the end.
+    ///         
+    /// Conditions for negative cases:
+    ///     - The containing method signature does not receive a ct parameter.
+    ///     - The invocation method signature receives a ct and one is already being explicitly passed, or...
+    ///     - The invocation method does not have an overload with the exact same arguments that also receives a ct, or...
+    ///     - The invocation method only has overloads that receive more than one ct.
+    /// </summary>
+    public abstract class ForwardCancellationTokenToInvocationsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2016";
+
+        protected abstract SyntaxNode? GetMethodNameNode(SyntaxNode invocationNode);
+
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsDescription),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
```

Comment:
[cosmetic] Most diagnostic use a non-split definition.

```suggestion
        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
```